design/logging.md

```diff
@@ -0,0 +1,495 @@
+/var/lib/gems/2.5.0/gems/actionpack-5.2.4.2/lib/action_dispatch/middleware/callbacks.rb:26:in `call'
+/var/lib/gems/2.5.0/gems/actionpack-5.2.4.2/lib/action_dispatch/middleware/debug_exceptions.rb:61:in `call'
+/var/lib/gems/2.5.0/gems/actionpack-5.2.4.2/lib/action_dispatch/middleware/show_exceptions.rb:33:in `call'
+/var/lib/gems/2.5.0/gems/railties-5.2.4.2/lib/rails/rack/logger.rb:38:in `call_app'
+/var/lib/gems/2.5.0/gems/railties-5.2.4.2/lib/rails/rack/logger.rb:26:in `block in call'
+/var/lib/gems/2.5.0/gems/activesupport-5.2.4.2/lib/active_support/tagged_logging.rb:71:in `block in tagged'
+/var/lib/gems/2.5.0/gems/activesupport-5.2.4.2/lib/active_support/tagged_logging.rb:28:in `tagged'
+/var/lib/gems/2.5.0/gems/activesupport-5.2.4.2/lib/active_support/tagged_logging.rb:71:in `tagged'
+/var/lib/gems/2.5.0/gems/railties-5.2.4.2/lib/rails/rack/logger.rb:26:in `call'
+/var/lib/gems/2.5.0/gems/actionpack-5.2.4.2/lib/action_dispatch/middleware/remote_ip.rb:81:in `call'
+/var/lib/gems/2.5.0/gems/actionpack-5.2.4.2/lib/action_dispatch/middleware/request_id.rb:27:in `call'
+/var/lib/gems/2.5.0/gems/rack-2.2.2/lib/rack/method_override.rb:24:in `call'
+/var/lib/gems/2.5.0/gems/rack-2.2.2/lib/rack/runtime.rb:22:in `call'
+/var/lib/gems/2.5.0/gems/activesupport-5.2.4.2/lib/active_support/cache/strategy/local_cache_middleware.rb:29:in `call'
+/var/lib/gems/2.5.0/gems/actionpack-5.2.4.2/lib/action_dispatch/middleware/executor.rb:14:in `call'
+/var/lib/gems/2.5.0/gems/actionpack-5.2.4.2/lib/action_dispatch/middleware/static.rb:127:in `call'
+/var/lib/gems/2.5.0/gems/rack-2.2.2/lib/rack/sendfile.rb:110:in `call'
+/var/lib/gems/2.5.0/gems/conjur-rack-heartbeat-2.2.0/lib/rack/heartbeat.rb:20:in `call'
+/var/lib/gems/2.5.0/gems/railties-5.2.4.2/lib/rails/engine.rb:524:in `call'
+/var/lib/gems/2.5.0/gems/puma-3.12.4/lib/puma/configuration.rb:227:in `call'
+/var/lib/gems/2.5.0/gems/puma-3.12.4/lib/puma/server.rb:675:in `handle_request'
+/var/lib/gems/2.5.0/gems/puma-3.12.4/lib/puma/server.rb:476:in `process_client'
+/var/lib/gems/2.5.0/gems/puma-3.12.4/lib/puma/server.rb:334:in `block in run'
+/var/lib/gems/2.5.0/gems/puma-3.12.4/lib/puma/thread_pool.rb:135:in `block in spawn_thread'
+ INFO 2020/05/18 09:45:41 +0000 [pid=298] [origin=127.0.0.1] [request_id=b7390d52-73a1-41ff-add3-9f4e37395e55] [tid=305] Completed 401 Unauthorized in 288ms
+```
+
+As you can see, this log shows too much data on why the request failed.
+This makes the log very hard to debug, and it's clear that we need some solution for this.
+
+## Solution
+
+We will redefine our log levels to the following:
+1. DEBUG - messages that can help Conjur developers investigate an issue. These messages will include
+           stack traces and DB messages.
+2. INFO - messages that can help Conjur users and operators investigate an issue. These messages will include
+          the messages that we print in our logs (messages with the `CONJ` prefix).
```

Comment:
I think that messages with prefix 'CONJ' with suffix 'E' need to be WARN (CONJ00007E for example). If i take your example:
`WARN 2020/05/17 17:39:59 +0000 [pid=401] [origin=127.0.0.1] [request_id=f39fb7e2-59f9-4f3a-89a9-bb9a50b09860] [tid=413] Authentication Error: #<Errors::Authentication::Security::RoleNotFound: CONJ00007E 'not_in_conjur' not found>`.

Reply:
I agree generally. However, this particular line has [a long history](https://github.com/cyberark/conjur/issues/1377) a needs approval from our security architects before we can this can show up in the Conjur logs by default (which is currently `WARN` level).

Reply:
the current default log level is `info`. I think that this message should be written in `INFO` (and the code should change accordingly). I address this issue in the doc.